ref(spool): Simplify the spool states

olksdr opened the pull request:
Simplify the state machine and have only 2 transition, from memory to disk and from disk to memory. 

Right now when the disk is full we start dropping the incoming envelopes and report the error.

In the followup will integrate the health check to report this issue to the user more gracefully. 

related: https://github.com/getsentry/team-ingest/issues/110

#skip-changelog

olksdr committed:
ref(spool): Simplify the spool states

jjbayer commented on relay-server/src/actors/spooler/mod.rs:
```diff
@@ -673,14 +644,14 @@ impl BufferService {
-            BufferState::Disk(ref disk) =>{
+            BufferState::Disk(ref disk) => {
+                // The disk is full, drop the incoming envelopes.
+                if disk.is_full().await? {
```
Idea for a potential follow-up: `is_full` is also called in `transition()` for every `handle_enqueue`, I wonder if we could cache it as a boolean flag on `OnDisk`, which gets updated on dequeue.

iker-barriocanal replied:
I'm not sure if I understand -- should `transition()` fail if we try to transition to `OnDisk` when the disk is full?

olksdr replied:
It will never happen, since we do not transition from Disk to Memory if the Disk has some data.

olksdr committed:
chore(spool): Integrate the health check into the spooler (#2113)

Once the disk is full, we will report the unhealthy to the client, to
signal that the Relay cannot accept any more envelopes.

olksdr committed:
Merge branch 'master' into chore/ref-buf-state
